OvmfPkg/IoMmuDxe/AmdSevIoMmu.c

```diff
@@ -513,6 +884,83 @@ IoMmuUnmap (
            );
 }
 
+/**
+ * Allocate CommonBuffer from pre-allocated shared memory.
+ *
+ * @param MemoryType          Memory type
+ * @param CommonBufferPages   Pages of CommonBuffer
+ * @param PhysicalAddress     Allocated physical address
+ * @param ReservedMemBitmap   Bitmap which indicates the allocation of reserved memory
+ * @return EFI_SUCCESS  Successfully allocate CommonBuffer
+ */
+STATIC
+EFI_STATUS
+AllocateCommonBuffer (
+  IN EFI_MEMORY_TYPE        MemoryType,
+  IN UINTN                  CommonBufferPages,
+  OUT EFI_PHYSICAL_ADDRESS  *PhysicalAddress,
+  OUT UINT32                *ReservedMemBitmap
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = InternalAllocateBuffer (
+             AllocateMaxAddress,
+             MemoryType,
+             CommonBufferPages,
+             ReservedMemBitmap,
+             PhysicalAddress
+             );
+  ASSERT (Status == EFI_SUCCESS);
+
+  mReservedMemBitmap |= *ReservedMemBitmap;
+
+  if (*ReservedMemBitmap != 0) {
+    *PhysicalAddress -= SIZE_4KB;
+  }
+
+  return Status;
+}
+
+/**
+ * Free CommonBuffer which is allocated by AllocateCommonBuffer().
+ *
+ * @param CommonBufferHeader  Pointer to the CommonBufferHeader
+ * @param CommonBufferPages   Pages of CommonBuffer
+ * @return EFI_SUCCESS        Successfully free the CommonBuffer

```

Comment:
It is still possble fail to FreePage, correct? is it covered at @retval EFI_FAILURE?

Reply:
You're right. It will be updated with @RetVal